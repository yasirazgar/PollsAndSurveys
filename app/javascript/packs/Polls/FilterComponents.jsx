import React,{ Fragment } from 'react';

import Select from '../Utils/Select'
import { AGE_SELECT_OPTIONS } from '../constants.js'

const FilterComponents = props => {
  let categories = props.categories.slice(0)
  categories.unshift([0, 'CATEGORIES'])
  let ageGroups = AGE_SELECT_OPTIONS

  return (
    <Fragment>
      <Select multiple={true} classes="categories" options={categories} onChange={props.categoryChangeHandler}/>

      <Select multiple={true} classes="agegroup" options={ageGroups} onChange={props.ageGroupChangeHandler}/>
    </Fragment>
  );
}

export default FilterComponents;
