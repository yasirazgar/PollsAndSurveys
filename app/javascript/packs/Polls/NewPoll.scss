.poll-question {
  background-color: green;
  padding: 30px 40px;
  color: white;
  text-align: center;
  input {
    margin: 5px;
    border: none;
    border-radius: 0;
    width: 75%;
    padding: 10px;
    float: left;
    font-size: 16px;
  }
  select {
    margin: 5px 0px;
    border: none;
    border-radius: 0;
    width: 30%;
    padding: 10px;
    font-size: 16px;
    &.categories {
      float: left;
    }
    &.agegroup {
      margin-right: 23%;
      float: right;
    }
  }
}

/* Clear floats after the poll__question */
.poll-question:after {
  content: "";
  display: table;
  clear: both;
}

/* Style the poll-question */
.poll-location {
  background-color: green;
  padding: 30px 40px;
  color: white;
  text-align: center;
  input {
    margin: 0;
    border: none;
    border-radius: 0;
    width: 75%;
    padding: 10px;
    float: left;
    font-size: 16px;
  }
}

.poll-option__new {
  padding: 10px;
  margin: -10px;
  width: 75%;
  font-size: 16px;
}