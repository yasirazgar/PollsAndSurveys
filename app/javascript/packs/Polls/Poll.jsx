import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Input from '../Utils/Input'

import './Poll.scss'


// Analyse whether this should be functional and class components
class Poll extends Component {
  answerPoll = () => {
    let alreadySelected = event.target.parentElement.getElementsByClassName("checked");
    if (alreadySelected.length > 0) {
      alreadySelected[0].classList.remove("checked");
    }
    event.target.classList.add('checked');
  }

  optionsList = () => {
    let options = []

    for (let i = 0; i < this.props.options.length; i++) {
      options.push(<li key={i.toString()} onClick={this.answerPoll}>{this.props.options[i]} </li>)
    }
    return options
  }

  render() {

    return (
      <div className="poll">
        <div className="poll__question">
          <h2>{this.props.question}</h2>
        </div>

        <ul className="poll__options">
          {this.optionsList()}
        </ul>
      </div>

    );
  }
}

Poll.propTypes = {
  question: PropTypes.string,
  options: PropTypes.array,
  categories: PropTypes.array
}

export default Poll
