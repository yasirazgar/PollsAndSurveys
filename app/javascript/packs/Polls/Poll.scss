.poll {
  margin-top: 50px;
  box-sizing: border-box;

  ul {
    margin: 0;
    padding: 0;

    /* Style the list items */
    li {
      cursor: pointer;
      position: relative;
      padding: 12px 8px 12px 40px;
      background: #eee !important;
      font-size: 18px;
      transition: 0.2s;

      /* make the list items unselectable */
      -webkit-user-select: none !important;
      -moz-user-select: none !important;
      -ms-user-select: none !important;
      user-select: none !important;
      list-style: none;
    }

    /* Set all odd list items to a different color (zebra-stripes) */
    li:nth-child(odd) {
      background: #f9f9f9 !important;
    }

    /* Darker background-color on hover */
    li:hover {
      background: #ddd;
    }

    /* When clicked on, add a background color and strike out text */
    li.checked {
      background-color: green !important;
      color: black;
    }

    /* Add a "checked" mark when clicked on */
    li.checked::before {
      content: '';
      position: absolute;
      border-color: #fff;
      border-style: solid;
      border-width: 0 2px 2px 0;
      top: 10px;
      left: 16px;
      transform: rotate(45deg);
      height: 15px;
      width: 7px;
    }
  }


  /* Style the close button */
  .close, .add {
    position: absolute;
    right: 0;
    top: 0;
    padding: 12px 16px 12px 16px;

    &:hover {
      background-color: #f44336;
      color: white;
    }
  }
}
