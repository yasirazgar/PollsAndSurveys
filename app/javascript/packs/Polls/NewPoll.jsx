import React, { Component } from 'react';
import { connect } from 'react-redux';

import Input from '../Utils/Input'
import Select from '../Utils/Select'
import ModalButtonGroup from '../Utils/ModalButtonGroup'
import FilterComponents from './FilterComponents'
import createPollHandler from '../Handlers/createPollHandler'

import './NewPoll.scss'

class NewPoll extends Component {
  constructor(props) {
    super(props);

    this.state = {
      question: '',
      options: [],
      category_ids: [],
      age_group_ids: [],
      submitEnabled: false,
    };

    this.enableSubmitButton = this.enableSubmitButton.bind(this);
    this.categoryChangeHandler = this.categoryChangeHandler.bind(this);
    this.setPollHandler = this.setPollHandler.bind(this);
  }

  categoryChangeHandler = (event) => {
    let selected_values = [...event.target.options].reduce((vals, opt) => {
      if(opt.selected){
        vals.push(opt.value)
      }; return vals
    }, [])

    this.setState({category_ids: selected_values})
    this.setPollHandler({
      question: this.state.question,
      options: this.state.options,
      category_ids: selected_values,
      age_group_ids: this.state.age_group_ids
    })
  }

  ageGroupChangeHandler = (event) => {
    let selected_values = [...event.target.options].reduce((vals, opt) => {
      if(opt.selected){
        vals.push(opt.value)
      }; return vals
    }, [])

    this.setState({age_group_ids: selected_values})
    this.setPollHandler({
      question: this.state.question,
      options: this.state.options,
      category_ids: this.state.category_ids,
      age_group_ids: selected_values
    })
  }

  setPollHandler = (poll) => {
    this.createPollHandler = createPollHandler.bind(this, poll, this.props.hideCreatePollForm)
  }

  enableSubmitButton = (question, options) => {
    let poll = {}
    if(question && options.length > 0){
      poll = {
        question: question,
        options: options,
        category_ids: this.state.category_ids,
        age_group_ids: this.state.age_group_ids
      }
      this.setState({submitEnabled: true});
    }
    else {
      this.setState({submitEnabled: false});
    }
    this.setPollHandler(poll);
  }

  buildOptionsList = () => {
    let optionsList = []

    for (let i = 0; i < this.state.options.length; i++) {
      optionsList.push(<li key={i.toString()} >{this.state.options[i]} <span className="close" onClick={this.removeOption.bind(this, i)}>Ã—</span></li>)
    }
    optionsList.push(<li key={this.state.options.length.toString()}> <input className="poll-option__new" placeholder={this.props.translations.options_placeholder} /> <span className="add" onClick={this.addOption.bind(this)}>+</span></li>)
    return optionsList;
  };

  setQuestion = (event) => {
    let question = event.target.value
    this.setState({question: question})
    // this.props.setPoll(question, this.state.options)
    this.enableSubmitButton(question, this.state.options);
  }

  removeOption = (i) => {
    let options = [...this.state.options];
    options.splice(i, 1);

    this.setState({options: options});
    this.enableSubmitButton(this.state.question, options);
  };

  addOption = () => {
    let options = this.state.options;
    let option = event.target.parentNode.getElementsByClassName('poll-option__new')[0].value;
    options.push(option);
    this.setState({options: options});
    this.enableSubmitButton(this.state.question, options);
    // this.props.setPoll(this.state.question, options)
  }

  render() {
    return (
      <div className="poll" id="new-poll">
        <div className="poll-question">
          <Input type="text" className="poll-question-input" placeholder={this.props.translations.question_placeholder} onChange={this.setQuestion}/>

          {/* <Input className="poll-location" placeholder="Location" /> */}
          <FilterComponents categories={this.props.categories} categoryChangeHandler={this.categoryChangeHandler} ageGroupChangeHandler={this.ageGroupChangeHandler} />
        </div>

        <ul >
          {this.buildOptionsList()}
        </ul>

        <ModalButtonGroup primaryText={this.props.translations.create} submitHandler={this.createPollHandler} closeModalHandler={this.props.hideCreatePollForm} submitEnabled={this.state.submitEnabled}/>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    translations: state.translations
  }
}

export default connect(mapStateToProps)(NewPoll)
