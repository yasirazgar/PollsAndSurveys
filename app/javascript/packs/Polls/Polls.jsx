import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchPolls } from '../../actions'

import Poll from './Poll'

// Analyse whether this should be functional and class components
class Polls extends Component {

  buildPolls = () => {
    const polls = this.props.polls.map((poll, i) => {
      return (<Poll isWithAnswer={false} key={i} poll={poll}/>)
    });
    return polls
  }

  componentDidMount() {
    this.props.fetchPolls();
  }

  render() {
    return this.buildPolls();
  }
}

const mapStateToProps = state => {
  return { polls: state.polls };
};

export default connect(mapStateToProps, { fetchPolls })(Polls)
