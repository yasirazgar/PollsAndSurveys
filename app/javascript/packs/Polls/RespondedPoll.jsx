import React, { Component  } from 'react';
import PropTypes from 'prop-types';

import Input from '../Utils/Input'
import { optionsListWithAnswer } from '../Helpers/polls_helper'

import './Poll.scss'

// Analyse whether this should be functional and class components
class RespondedPoll extends Component {
  constructor(props){
    super(props)
    this.state = {
      poll: this.props.poll,
    }
    this.setAnswers = this.setAnswers.bind(this);
    this.buildOptionsList = this.buildOptionsList.bind(this);
  }

  setAnswers = (poll) => {
    this.setState({poll: poll})
  }

  buildOptionsList = () => optionsListWithAnswer(this.state.poll, this.setAnswers);

  render() {
    return (
      <div className="poll">
        <div className="poll__question">
          <h2>{this.state.poll.question}</h2>
        </div>

        <ul className="poll__options">
          {this.buildOptionsList()}
        </ul>
      </div>
    );
  }
}

RespondedPoll.propTypes = {
  question: PropTypes.string,
  options: PropTypes.array,
  categories: PropTypes.array
}

export default RespondedPoll
