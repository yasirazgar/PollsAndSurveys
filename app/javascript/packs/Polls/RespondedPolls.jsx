import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchRespondedPolls } from '../../actions'

import Poll from './Poll'

// Analyse whether this should be functional and class components
class RespondedPolls extends Component {

  buildUserPolls = () => {
    let pollsList = this.props.polls.map((poll, i) => {
      return (<Poll isWithAnswer={true} key={i} poll={poll} />)
    });
    return pollsList;
  }

  componentDidMount() {
    this.props.fetchRespondedPolls()
  }

  render() {
    return this.buildUserPolls();
  }
}

const mapStateToProps = state => {
  return { polls: state.respondedPolls };
};

export default connect(mapStateToProps, { fetchRespondedPolls })(RespondedPolls)

