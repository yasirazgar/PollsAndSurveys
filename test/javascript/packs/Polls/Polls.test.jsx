import React from 'react'
import { shallow } from 'enzyme'
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

import Polls from 'packs/Polls/Polls'

const mockStore = configureMockStore();

describe("Polls", () => {
  const polls = [
    {
      poll_id: 1,
      question: 'Fav snake',
      categories: [[3, 'Animals']],
      options: {
        Python: {option_id: 1, percentage: 50.0, selected: true},
        Cobra: {option_id: 2, percentage: 25.0, selected: false},
        Viper: {option_id: 3, percentage: 25.0, selected: false},
        Mamba: {option_id: 4, percentage: 0.0,  selected: false}
      }
    },
    {
      poll_id: 1,
      question: 'Fav snake',
      categories: [[3, 'Animals']],
      options: {
        Python: {option_id: 1, percentage: 50.0, selected: true},
        Cobra: {option_id: 2, percentage: 25.0, selected: false},
        Viper: {option_id: 3, percentage: 25.0, selected: false},
        Mamba: {option_id: 4, percentage: 0.0,  selected: false}
      }
    }
  ]

  it("renders correctly with search poll", () => {
    const wrapper = shallow(
      <Provider store={mockStore({})}>
        <Polls searchPoll={polls} />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly with polls", () => {
    const wrapper = shallow(
      <Provider store={mockStore({})}>
        <Polls polls={polls} />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly with polls", () => {
    const wrapper = shallow(
      <Provider store={mockStore({})}>
        <Polls userPolls={polls} />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly with respondedPolls", () => {
    const wrapper = shallow(
      <Provider store={mockStore({})}>
        <Polls respondedPolls={polls} />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot();
  });
})
