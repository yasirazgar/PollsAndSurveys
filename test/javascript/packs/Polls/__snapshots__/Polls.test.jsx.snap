// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polls renders correctly with polls 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <Connect(Polls)
    polls={
      Array [
        Object {
          "categories": Array [
            Array [
              3,
              "Animals",
            ],
          ],
          "options": Object {
            "Cobra": Object {
              "option_id": 2,
              "percentage": 25,
              "selected": false,
            },
            "Mamba": Object {
              "option_id": 4,
              "percentage": 0,
              "selected": false,
            },
            "Python": Object {
              "option_id": 1,
              "percentage": 50,
              "selected": true,
            },
            "Viper": Object {
              "option_id": 3,
              "percentage": 25,
              "selected": false,
            },
          },
          "poll_id": 1,
          "question": "Fav snake",
        },
        Object {
          "categories": Array [
            Array [
              3,
              "Animals",
            ],
          ],
          "options": Object {
            "Cobra": Object {
              "option_id": 2,
              "percentage": 25,
              "selected": false,
            },
            "Mamba": Object {
              "option_id": 4,
              "percentage": 0,
              "selected": false,
            },
            "Python": Object {
              "option_id": 1,
              "percentage": 50,
              "selected": true,
            },
            "Viper": Object {
              "option_id": 3,
              "percentage": 25,
              "selected": false,
            },
          },
          "poll_id": 1,
          "question": "Fav snake",
        },
      ]
    }
  />
</ContextProvider>
`;

exports[`Polls renders correctly with polls 2`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <Connect(Polls)
    userPolls={
      Array [
        Object {
          "categories": Array [
            Array [
              3,
              "Animals",
            ],
          ],
          "options": Object {
            "Cobra": Object {
              "option_id": 2,
              "percentage": 25,
              "selected": false,
            },
            "Mamba": Object {
              "option_id": 4,
              "percentage": 0,
              "selected": false,
            },
            "Python": Object {
              "option_id": 1,
              "percentage": 50,
              "selected": true,
            },
            "Viper": Object {
              "option_id": 3,
              "percentage": 25,
              "selected": false,
            },
          },
          "poll_id": 1,
          "question": "Fav snake",
        },
        Object {
          "categories": Array [
            Array [
              3,
              "Animals",
            ],
          ],
          "options": Object {
            "Cobra": Object {
              "option_id": 2,
              "percentage": 25,
              "selected": false,
            },
            "Mamba": Object {
              "option_id": 4,
              "percentage": 0,
              "selected": false,
            },
            "Python": Object {
              "option_id": 1,
              "percentage": 50,
              "selected": true,
            },
            "Viper": Object {
              "option_id": 3,
              "percentage": 25,
              "selected": false,
            },
          },
          "poll_id": 1,
          "question": "Fav snake",
        },
      ]
    }
  />
</ContextProvider>
`;

exports[`Polls renders correctly with respondedPolls 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <Connect(Polls)
    respondedPolls={
      Array [
        Object {
          "categories": Array [
            Array [
              3,
              "Animals",
            ],
          ],
          "options": Object {
            "Cobra": Object {
              "option_id": 2,
              "percentage": 25,
              "selected": false,
            },
            "Mamba": Object {
              "option_id": 4,
              "percentage": 0,
              "selected": false,
            },
            "Python": Object {
              "option_id": 1,
              "percentage": 50,
              "selected": true,
            },
            "Viper": Object {
              "option_id": 3,
              "percentage": 25,
              "selected": false,
            },
          },
          "poll_id": 1,
          "question": "Fav snake",
        },
        Object {
          "categories": Array [
            Array [
              3,
              "Animals",
            ],
          ],
          "options": Object {
            "Cobra": Object {
              "option_id": 2,
              "percentage": 25,
              "selected": false,
            },
            "Mamba": Object {
              "option_id": 4,
              "percentage": 0,
              "selected": false,
            },
            "Python": Object {
              "option_id": 1,
              "percentage": 50,
              "selected": true,
            },
            "Viper": Object {
              "option_id": 3,
              "percentage": 25,
              "selected": false,
            },
          },
          "poll_id": 1,
          "question": "Fav snake",
        },
      ]
    }
  />
</ContextProvider>
`;

exports[`Polls renders correctly with search poll 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <Connect(Polls)
    searchPoll={
      Array [
        Object {
          "categories": Array [
            Array [
              3,
              "Animals",
            ],
          ],
          "options": Object {
            "Cobra": Object {
              "option_id": 2,
              "percentage": 25,
              "selected": false,
            },
            "Mamba": Object {
              "option_id": 4,
              "percentage": 0,
              "selected": false,
            },
            "Python": Object {
              "option_id": 1,
              "percentage": 50,
              "selected": true,
            },
            "Viper": Object {
              "option_id": 3,
              "percentage": 25,
              "selected": false,
            },
          },
          "poll_id": 1,
          "question": "Fav snake",
        },
        Object {
          "categories": Array [
            Array [
              3,
              "Animals",
            ],
          ],
          "options": Object {
            "Cobra": Object {
              "option_id": 2,
              "percentage": 25,
              "selected": false,
            },
            "Mamba": Object {
              "option_id": 4,
              "percentage": 0,
              "selected": false,
            },
            "Python": Object {
              "option_id": 1,
              "percentage": 50,
              "selected": true,
            },
            "Viper": Object {
              "option_id": 3,
              "percentage": 25,
              "selected": false,
            },
          },
          "poll_id": 1,
          "question": "Fav snake",
        },
      ]
    }
  />
</ContextProvider>
`;
